enhancements/ingress/configurable-dns-loadbalancerservice.md

```diff
@@ -0,0 +1,316 @@
+the cluster admin. 
+
+#### Modifying/Updating an ingresscontroller
+
+The ingresscontroller will need to be deleted and recreated with the updated domain
+present in if required domain is _not_ the same as cluster domain configured at
+*ingress.config.openshift.io/cluster* `.spec.domain`. If not needed, the workflow
+defined below is sufficient,
+- The new domain that needs to be associated with the ingresscontroller
+    must be created prior to making any changes.
+- The ingresscontroller must be edited to set `.loadBalancer.dns` to `Unmanaged`.
+- This will trigger a reconcile of the controller, resulting in updating
+    the following conditions on the ingress operator
+  - `DNSManaged` condition will be set to false and reason updated to
+    UnmanagedDNS.
+- The DNSRecord will also be updated as part of the same reconcile, 
+  `.spec.managementPolicy` will be set to `Unmanaged` and the following conditions
+  will be updated,
+  - `DNSReady` condition will be set to true and reason updated to
+    UnmanagedDNS.
+   
+    
+__Note__: Similarly, to create a new ingresscontroller where DNS is managed
+          externally, the same workflow can be followed. No *DNSRecord*s will be
+          created in the flow.
+
+### API Extensions
+
+The ingresscontroller CRD is extended by adding an optional field `DNS`
+of type `LoadBalancerDNSManagementPolicy`. This field will default to `Managed`.
+
+```go
+// LoadBalancerStrategy holds parameters for a load balancer.
+type LoadBalancerStrategy struct {
+    // <snip>
+
+    // dns indicates if the lifecyle of the wildcard DNS record
+    // associated with the load balancer service will be managed by
+    // the ingress operator.
+    //
+    // +kubebuilder:default:="Managed"
+    // +kubebuilder:validation:Optional
+    // +kubebuilder:validation:Enum=Managed;Unmanaged
+    // +optional
+    DNS LoadBalancerDNSManagementPolicy `json:"dns"`
+}
+
+// LoadBalancerDNSManagementPolicy is a policy for configuring how
+// ingresscontrollers manage DNS.
+//
+// +kubebuilder:validation:Enum=Managed;Unmanaged
+type LoadBalancerDNSManagementPolicy string
+
+const (
+    // ManagedLoadBalancerDNS configures the ingresscontroller to manage
+    // the wildcard DNS records associated with it.
+    ManagedLoadBalancerDNS LoadBalancerDNSManagementPolicy = "Managed"
+    // UnmanagedLoadBalancerDNS configures the ingresscontroller to not
+    // manage the wildcard DNS records associated with it.
+    UnmanagedLoadBalancerDNS LoadBalancerDNSManagementPolicy = "Unmanaged"
+)
+
+// DNSRecordSpec contains the details of a DNS record.
+type DNSRecordSpec struct {
+  // <snip>
+
+  // managementPolicy denotes the current policy applied on the dns record.
+  // Records that have policy set as "Unmanaged" are ignored by the ingress
+  // operator and should be managed at the discretion of the cluster admin.
+  //
+  // +kubebuilder:default:="Managed"
+  // +kubebuilder:validation:Optional
```

Comment:
Sharing a bit of feedback I've gotten from the API team on this: https://github.com/openshift/api/pull/1110#discussion_r836645159

Reply:
I believe that the idea of using both `+optional` and `+kubebuilder:validation:Optional` comes from the support of different code generators. `+kubebuilder:*` markers work only for kubebuilder while `+optional` is a universal convention which can enable the generation done by a different code generator.

Reply:
1. No, I don't think so.
2. No. Yes, I agree that the kubebuilder specific marker is preferred for the code generation.

When I saw them both specified I was also wondering whether this redundancy is really needed, however I thought that keeping them would still be needed because:

- it's consistent with the other OpenShift APIs, [operator's ingress one too](https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go) (even though just `+optional` is used in some places).
- it's part of [Kube API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) (OpenAPI documentation is mentioned as a tool which may need it)

Reply:
ack, understood. I think it may be worth getting some more formal guidance on this from the API team. With that said, we can close this thread and let the API team decide what needs to happen (if anything). Thanks @alebedev87 !